### ISSUE TITLE
Extract the models into a library?

### ISSUE THREAD

**Issue by aSapien** [slack-scala-client/slack-scala-client#101]
First, very impressive work on modeling the API entities of slack!

I would like to be able to depend on the `models` package and create a [circe](https://github.com/circe/circe) parser with any relevant encoders/decoders. This means that any Play code needs to be extracted out of the `models` package and the package published under a separate artifact. 

Does this sound like a reasonable idea?

I tried to come up with an automatic `case class` generator from [their OpenAPI spec](https://api.slack.com/web#spec) [(Download link)](https://api.slack.com/specs/openapi/v2/slack_web.json) but I failed to find a tool that would succeed in parsing their spec without generating tons of errors:
- I tried to use [sbt-swagger-codegen](https://github.com/unicredit/sbt-swagger-codegen) but it failed with `Invalid swagger format` error. 
- Also attempted to use [openapi2proto](https://github.com/nytimes/openapi2proto) and also failed with some unsupported format error. 

Please advise

**Comment by mccartney** [comment]
That's a good point @aSapien. I haven't found any smart way to generate these models either. And I bet they were created manually by the original library author.

As to your request - what would be the benefit of publishing the models separately. Wouldn't it just possible for you to include the whole library and use only the single package? It's not like the rest is huge.

**Comment by aSapien** [comment]
@mccartney IMO the benefits of a separate package is a much lighter deptree and less likelihood of dependency version collisions.

**Comment by mccartney** [comment]
OK. I get your point. My only worry is the excessive release (and related) work with this. I'd need to research whether to model that as two separate projects or some modules (?). I don't have open-source experience with such setups with `sbt`.

Any hints from anyone?

**Comment by mccartney** [comment]
I was thinking more about your suggestion. Would like the "models only" sublibrary/project contain only the case-classes or also the Json deserialization/serialization code?
The latter would come with dependencies.

(BTW, I am thinking of changing the Json handling code here. As it is now, it comes with lot of repetition and boilerplate. I am not an expert in Json handling in Scala, but I bet there's some simpler way)

**Comment by tg44** [comment]
Do you want somebody to work on this? I'm generally familiar with sbt multiprojects, only the publishing is the "grey area".

**Comment by mccartney** [comment]
We are happy to take contributions here!
Being realistic I don't think I will have time to do it myself.

So go ahead! I offer to be responsive when pinged for particular issues.

**Comment by mccartney** [comment]
Related to #13

**Comment by mccartney** [comment]
I am not fully sure if I fully get your question, but yes - it looks like `InvalidResponseError` and `ApiError` do not belong to what be the extracted models submodule/subproject. Despite they are currently in the `slack.models` package. These two (and maybe others) should be kept with the API code I think.

**Comment by mfirry** [comment]
Alright but this creates somehow a dependency cycle no? with `models` depending on `api` and vice versa.

**Comment by mfirry** [comment]
Right. 
So potentially `RetryAfter` can be moved to its own file on `api`? Are you okay with that? 

The rest of `modules` seems ok on its own and also "almost" with zero dependencies (I mean possibly just play-json).

**Comment by mccartney** [comment]
>  So potentially RetryAfter can be moved to its own file on api? Are you okay with that?

Yes

>  (I mean possibly just play-json).

I think the goal was not to have play-json in the models library either :D

**Comment by mfirry** [comment]
Ok talking about the dependency from play-json that models has... I think I need your help. For example within `Block` the `TextObject` object relies a lot on play-json (which is fine) but I think there's need here to define a strategy _if_ we want to separate `models` and make it sorf of independent.

A few options:
- move JSON encoding/deconding in its own module (something like `slack-scala-client-models-play-json`)
- move dependency from play-json on the client itself (and JSON encoding/deconding there as well)
- a _typeclass_ sort of approach in which you define a `JSONable` trait in the models/api and then require implementations so potentially you can have play-json, circe or spray-json 

What do you say?

**Comment by mccartney** [comment]
Good question. I don't have a good answer (yet?).

Thinking out loud...

Indeed the `Reads` and `Writes` are Play-json specific way of providing serialization of these objects. The fact it's using Play is our choice. The shape/configuration of that serialization is actually **modelling** the external design by Slack. So maybe the models should contain both:
- the modelling of the notions
- and how they can be serialized/deserialized to JSON?

But then we would drag the dependency on Play to models, which is literally against the original idea of this issue.

So looking at your three options. Three modules is too much I think. I mean I don't see a value for having these three separate: models, serialization, API interfacing. The typeclass kind of approach is smart, but I think it's bothersome - we would need to maintain ~3 implementations of these serializations (for what currently seems little benefit).

So I think I am in favor of "move dependency from play-json on the client itself". But I am happy to be pointed to some other arguments I didn't see.

**Comment by tg44** [comment]
Do we really need to choose/force/enable multiple serializations? If our api is closed enough, the inner implementation doesn't matter, and we could even parse the incoming jsons with monkeys. It is just a transitive dependency, that nobody cares about.

I mean if the end-users api is a `def doSomething(a: Model1, b: Model2): Model3` like, I really don't care if it is serialized with circe or spray or argonaut or some other libs, I just want to call the `doSomething`.

As I started to work on this a year ago, I quickly found out that the requesting interface holds the bigger challenge, this is why I moved it to taglessFinal at first, so we could change more freely both the requester (akka-http) and the serializer (spray) implementations. The discussion stopped after a while, but I think that is definitely a good direction.

If you check my refactor direction, I moved the generalization into the `SlackApiClientF`, and kept its current interfaces. If we move out the requester from the `SlackApiClient`, we could write a pure tagless final implementation, which is only dependant on the requester and some basic helper functions on the F we want to use and a small wrapper for the json lib we want to use (like map, json.as (of any type of json libs), etc.). So a http4s + circe impementation would be simply adding a http4s requester instance and a circe json parser instance. We could easily change out the F too, so it could use ZIO, IO, Future, Id(/Try?). I didn't go deeper into this direction bcs I wanted to get more feedback/approval, but I'm happy that others are interested in this too :)
